src/vopr.zig

```diff
@@ -681,34 +681,47 @@ pub const Simulator = struct {
+            if (simulator.core.isSet(replica.replica)) {
+                cluster_op_checkpoint = @max(
+                    cluster_op_checkpoint,
+                    replica.superblock.working.vsr_state.checkpoint.header.op,
+                );
+            }
+        }
 
         var missing_header_op: ?u64 = null;
         var missing_prepare_op: ?u64 = null;
 
         for (simulator.cluster.replicas) |replica| {
             if (simulator.core.isSet(replica.replica) and !replica.standby()) {
                 assert(simulator.cluster.replica_health[replica.replica] == .up);
+
+                // Lagging replicas do not initiate WAL repair during view change.
+                if (replica.status == .view_change and
+                    replica.op_checkpoint() < cluster_op_checkpoint) continue;
+
                 const commit_max = @min(replica.op, cluster_commit_max);
-                const commit_min = replica.commit_min;
-                if (commit_min < commit_max) {
-                    // Check if replica was stuck while repairing headers. Find largest missing
-                    // header as we repair headers from high -> low ops.
-                    if (replica.journal.find_latest_headers_break_between(
-                        commit_min + 1,
-                        commit_max,
-                    )) |range| {
-                        if (missing_header_op == null or missing_header_op.? < range.op_max) {
-                            missing_header_op = range.op_max;
-                        }
-                    } else {
-                        // Check if replica was stuck while repairing prepares. Find smallest
-                        // missing prepare as we repair prepares from low -> high ops.
-                        for (commit_min + 1..commit_max + 1) |op| {
-                            const header = simulator.cluster.state_checker.header_with_op(op);
-                            if (!replica.journal.has_clean(&header)) {
-                                if (missing_prepare_op == null or missing_prepare_op.? > op) {
-                                    missing_prepare_op = op;
-                                }
+                const op_repair_min = replica.op_repair_min();
+
+                // Check if replica's commit pipeline was stuck due to missing headers.
+                // Find largest missing header as we repair headers from high -> low ops.
+                if (replica.journal.find_latest_headers_break_between(
+                    op_repair_min,
+                    commit_max,
+                )) |range| {
+                    if (missing_header_op == null or missing_header_op.? < range.op_max) {
+                        missing_header_op = range.op_max;
+                    }
+                } else {
+                    // Check if replica's commit pipeline was stuck due to missing prepares.
+                    // Find smallest missing prepare as we repair prepares from low -> high ops.
```

Comment:
Nit, but I would clarify s/smallest/earliest, since the size of the prepare is irrelevant.

Reply:
Ah, nice catch there! I did mean earliest & latest, but used smallest & largest; was perhaps thinking in terms of smallest index & largest index. 😄 

Addressed in https://github.com/tigerbeetle/tigerbeetle/pull/2409/commits/2fc1f816e89365b164f37ce384c115474a4dd867.